Add support for slice, selec twith int, index_select

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64323 Add embedding shape analysis
* #64121 Max Pool with indices
* #63530 Add Maxpool to shape analysis / Opinfo
* #64300 Add symbolic shape comparison optimization
* #64299 Refactor to use shape arguments
* #63516 Add view with negative dim
* #63615 Generalize expand logic
* #63407 Add permute, arange
* **#63365 Add support for slice, selec twith int, index_select**
* #63099 Add squeeze, unsqueeze, transpose shape functins
* #63050 Add batch of unary functions

Differential Revision: [D30738144](https://our.internmc.facebook.com/intern/diff/D30738144)

Commit:
Add support for slice, selec twith int, index_select

[ghstack-poisoned]

Commit:
Update on "Add support for slice, selec twith int, index_select"




[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -516,6 +516,7 @@ def __init__(self,
                  # the following metadata relates to sparse support and is used in test_sparse.py
                  supports_sparse=False,  # whether the op supports sparse inputs
 
+                 skip_jit_scripting=False,  # only run tracing tests
```
Should we model this directly as "supports_scripting"?

Typically skip metadata should be in a SkipInfo. That would work if the above test was split into tracing and scripting versions, which might make more sense than modeling a partial test skip as operator metdata.

eellison replied:
yea, i agree it would be nice to eventually split these tests up, but I would prefer to prioritize expanding op coverage first. Fine with `supports_scripting`. There are a lot of fixes I think are higher-pri than splitting them up, I have a WIP PR to improve autodiff for instance and there are a ton of skipped JIT tests for pretty small reasons.

mruberry replied:
That sounds good

eellison replied:
let me actually try to rope more people in to fixing this stuff, and maybe we can split up tracing and scripting sooner..

Commit:
Update on "Add support for slice, selec twith int, index_select"




[ghstack-poisoned]

Commit:
Update on "Add support for slice, selec twith int, index_select"




[ghstack-poisoned]

Commit:
Update on "Add support for slice, selec twith int, index_select"




[ghstack-poisoned]

makslevental commented on torch/csrc/jit/runtime/symbolic_shape_registry.cpp:
```diff
@@ -131,6 +131,62 @@ const std::string shape_compute_functions =
+        def index_select(self: List[int], dim: int, index: List[int]):
+          dim = maybe_wrap_dim(dim, len(self))
+          numel = multiply_integers(index)
+          assert len(index) <= 1
```
jw why pass in `List[int]` if it has to be basically an int? Is this necessitated by some public facing api somewhere?

eellison replied:
The operator signature is `aten::index_select(Tensor self, int dim, Tensor index)`, and here this is logic that is being replicated here

https://codebrowser.bddppq.com/pytorch/pytorch/aten/src/ATen/native/TensorAdvancedIndexing.cpp.html#858 -

makslevental commented on torch/csrc/jit/runtime/symbolic_shape_registry.cpp:
```diff
@@ -131,6 +131,62 @@ const std::string shape_compute_functions =
               out.append(li[i])
           return out
 
+        def index_select(self: List[int], dim: int, index: List[int]):
```
nit?: should this have a return type `-> int`?

eellison replied:
return type gets inferred

makslevental commented on torch/csrc/jit/runtime/symbolic_shape_registry.cpp:
```diff
@@ -131,6 +131,62 @@ const std::string shape_compute_functions =
+          return result_size
+
+        def max_int():
+          return 9223372036854775807
```
nit: i guess this is `sys.maxsize`. is there a reason you can't just use that? i can't tell where this python-esque stuff runs...

eellison replied:
it's not bound yet in torchscript.. i can add a TODO

makslevental reviewed (approved):
LGTM

Commit:
Update on "Add support for slice, selec twith int, index_select"




[ghstack-poisoned]

Commit:
Update on "Add support for slice, selec twith int, index_select"




[ghstack-poisoned]

Commit:
Update on "Add support for slice, selec twith int, index_select"




[ghstack-poisoned]

Commit:
Update on "Add support for slice, selec twith int, index_select"




[ghstack-poisoned]

Commit:
Update on "Add support for slice, selec twith int, index_select"




[ghstack-poisoned]

Commit:
Update on "Add support for slice, selec twith int, index_select"




[ghstack-poisoned]

Commit:
Update on "Add support for slice, selec twith int, index_select"




[ghstack-poisoned]

Commit:
Update on "Add support for slice, selec twith int, index_select"




[ghstack-poisoned]

Commit:
Update on "Add support for slice, selec twith int, index_select"




[ghstack-poisoned]

Commit:
Update on "Add support for slice, selec twith int, index_select"




[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30738144).

Commit:
Update on "Add support for slice, selec twith int, index_select"


Differential Revision: [D30738144](https://our.internmc.facebook.com/intern/diff/D30738144)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30738144).

Commit:
Update on "Add support for slice, selec twith int, index_select"


Differential Revision: [D30738144](https://our.internmc.facebook.com/intern/diff/D30738144)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30738144).
